### DISCUSSION TITLE
How to load separate scripts from plugin/init.lua

### DISCUSSION THREAD

**Discussion by oscarcederberg** [wezterm/wezterm#4150] [Q&A]
i'm attempting to create a simple plugin that loads my configurations from github. i've broken up my configuration into different scripts and folders, all loaded in by an `init.lua`-script.

alongside my scripts i also have an asset-folder, with pngs/gifs for a custom background.

```
plugin/
- assets/
- - background.png
- - etc...
- color.lua
- init.lua
- key.lua
- set.lua
```
my question is, how do i go about loading in these separate files in a wezterm plugin? how would i also go about **importing the assets** without knowing the absolute path of them?

i know i could put everything into one script, but i'd prefer to learn how to do this.

i've attempted the following:
- put them in `plugin/`, and load them with `require('<xyz>')`.
- same as above but instead load them with `require('plugin.<xyz>')`.
- put them in `plugin/knos/` alongside a `init.lua` that loads them and then loading that script with `require('knos')`.
- put them into a seperate folder `knos/` and do the same as above.
- put them into a seperate folder `lua/knos/`

most of the above were just desperate attempts based on the layout of neovim plugins. i haven't found any wezterm plugins that use several scripts, unfortunately.

thanks!

**Comment by oscarcederberg** [comment]
i think i found the answer from #3989 and https://github.com/MLFlexer/smart_workspace_switcher.wezterm/tree/main

in `plugin/init.lua`:
```lua
local wezterm = require("wezterm")
local module = {}

local function resolve_path()
  for _, plugin in ipairs(wezterm.plugin.list()) do
    if plugin.url == "<REPO_URL>" then
      return plugin.plugin_dir
    end
  end
end

function module.apply_to_config(config)
  local other_module
  local path = resolve_path()

  if not path then
    wezterm.log_error("<REPO_NAME> not found")
    return
  end

  -- adds `<PLUGIN_DIR>/lua/` as a "require-able" path
  package.path = path .. "/lua/?/init.lua;" .. package.path
  -- retrieves `<PLUGIN_DIR>/lua/other_module/init.lua`
  other_module = require("other_module")
  other_module.apply_to_config(config)
end

return module
```

i think it is possible to retrieve asset paths similarly. no idea if `other_module/init.lua` is gonna be able to call e.g. `require("other_module.color")` or similar scripts inside the same directory as itself. unfortunately `plugin_dir` is not released yet, so i can't test right now and see if it works. just gotta wait and see :smile:

**Reply by wez** [reply]
I'm not sure that changing the package.path is the best idea, as I think that there is a higher chance of collisions between plugins.

Can you try something like this more general lua technique and see how it works?

https://stackoverflow.com/a/9146653/149111

**Reply by oscarcederberg** [reply]
thanks! never found that script/library name could be retrieved with `...` while searching myself, nice!!

logging `...` returns the URL-encoded name of the plugin, and adding `.<NAME>` to it leads to it looking for `/run/user/1000/wezterm/plugins/<ENCODED_PLUGIN_NAME>/<NAME>/plugin/init.lua` (note the last `plugin/`-dir).

this method works OK for a single script, but becomes quite unwieldy for several scripts.

e.g. in `plugin/init.lua`:
```lua
local wezterm = require("wezterm")
local module = {}
local path = ...

function module.apply_to_config (config)
  local knos = require (path .. ".knos")
  local knos_colors = require (path .. ".knos.colors")
end

return module
```
would look for `knos/plugin/init.lua` and `knos/colors/plugin/init.lua`.

i could modify `package.path` to look in `/run/user/1000/wezterm/plugins/<ENCODED_PLUGIN_NAME>/<NAME>/init.lua` (not including the last `plugin/`), but the problem still remains with collisions, retrieval of e.g. `knos/colors.lua`, and is probably not portable between OSs.

i think i should just accept that i should just use a single script instead. for finding the path to assets i can simply wait until `plugin_dir` is available. 😄

**Reply by oscarcederberg** [reply]
I created a pull request regarding this #4248
